// @openzeppelin/hardhat-upgrades/dist/defender/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseTxOverrides = exports.waitForDeployment = exports.getRemoteDeployment = exports.getDeployClient = exports.disableDefender = exports.enableDefender = exports.getNetwork = exports.getDefenderApiKey = void 0;
const upgrades_core_1 = require("@openzeppelin/upgrades-core");
const defender_sdk_base_client_1 = require("@openzeppelin/defender-sdk-base-client");
const defender_sdk_deploy_client_1 = require("@openzeppelin/defender-sdk-deploy-client");
const debug_1 = __importDefault(require("../utils/debug"));
const util_1 = require("util");
const sleep = (0, util_1.promisify)(setTimeout);
function getDefenderApiKey(hre) {
    const cfg = hre.config.defender;
    if (!cfg || !cfg.apiKey || !cfg.apiSecret) {
        const sampleConfig = JSON.stringify({ apiKey: 'YOUR_API_KEY', apiSecret: 'YOUR_API_SECRET' }, null, 2);
        throw new Error(`Missing OpenZeppelin Defender API key and secret in hardhat config. Add the following to your hardhat.config.js configuration:\ndefender: ${sampleConfig}\n`);
    }
    return cfg;
}
exports.getDefenderApiKey = getDefenderApiKey;
async function getNetwork(hre) {
    const { provider } = hre.network;
    const chainId = hre.network.config.chainId ?? (await (0, upgrades_core_1.getChainId)(provider));
    const network = (0, defender_sdk_base_client_1.fromChainId)(chainId);
    if (network === undefined) {
        throw new Error(`Network ${chainId} is not supported by OpenZeppelin Defender`);
    }
    return network;
}
exports.getNetwork = getNetwork;
function enableDefender(hre, defenderModule, opts) {
    if ((hre.config.defender?.useDefenderDeploy || defenderModule) && opts.useDefenderDeploy === undefined) {
        return {
            ...opts,
            useDefenderDeploy: true,
        };
    }
    else {
        return opts;
    }
}
exports.enableDefender = enableDefender;
/**
 * Disables Defender for a function that does not support it.
 * If opts.useDefenderDeploy or defenderModule is true, throws an error.
 * If hre.config.defender.useDefenderDeploy is true, logs a debug message and passes (to allow fallback to Hardhat signer).
 *
 * @param hre The Hardhat runtime environment
 * @param defenderModule Whether the function was called from the defender module
 * @param opts The options passed to the function
 * @param unsupportedFunction The name of the function that does not support Defender
 */
function disableDefender(hre, defenderModule, opts, unsupportedFunction) {
    if (opts.useDefenderDeploy) {
        throw new upgrades_core_1.UpgradesError(`The function ${unsupportedFunction} is not supported with the \`useDefenderDeploy\` option.`);
    }
    else if (defenderModule) {
        throw new upgrades_core_1.UpgradesError(`The function ${unsupportedFunction} is not supported with the \`defender\` module.`, () => `Call the function as upgrades.${unsupportedFunction} to use the Hardhat signer.`);
    }
    else if (hre.config.defender?.useDefenderDeploy) {
        (0, debug_1.default)(`The function ${unsupportedFunction} is not supported with the \`defender.useDefenderDeploy\` configuration option. Using the Hardhat signer instead.`);
    }
}
exports.disableDefender = disableDefender;
function getDeployClient(hre) {
    return new defender_sdk_deploy_client_1.DeployClient(getDefenderApiKey(hre));
}
exports.getDeployClient = getDeployClient;
/**
 * Gets the remote deployment response for the given id.
 *
 * @param hre The Hardhat runtime environment
 * @param remoteDeploymentId The deployment id.
 * @returns The remote deployment response, or undefined if the deployment is not found.
 * @throws Error if the deployment response could not be retrieved.
 */
async function getRemoteDeployment(hre, remoteDeploymentId) {
    const client = getDeployClient(hre);
    try {
        return (await client.getDeployedContract(remoteDeploymentId));
    }
    catch (e) {
        const message = e.response?.data?.message;
        if (message?.match(/deployment with id .* not found\./)) {
            return undefined;
        }
        throw e;
    }
}
exports.getRemoteDeployment = getRemoteDeployment;
/**
 * Waits indefinitely for the deployment until it is completed or failed.
 * Returns the last known transaction hash seen from the remote deployment, or undefined if the remote deployment was not retrieved.
 */
async function waitForDeployment(hre, opts, address, remoteDeploymentId) {
    const pollInterval = opts.pollingInterval ?? 5e3;
    let lastKnownTxHash;
    // eslint-disable-next-line no-constant-condition
    while (true) {
        if (await (0, upgrades_core_1.hasCode)(hre.ethers.provider, address)) {
            (0, debug_1.default)('code in target address found', address);
            break;
        }
        const response = await getRemoteDeployment(hre, remoteDeploymentId);
        lastKnownTxHash = response?.txHash;
        const completed = await (0, upgrades_core_1.isDeploymentCompleted)(address, remoteDeploymentId, response);
        if (completed) {
            break;
        }
        else {
            await sleep(pollInterval);
        }
    }
    return lastKnownTxHash;
}
exports.waitForDeployment = waitForDeployment;
function parseTxOverrides(overrides) {
    if (!overrides) {
        return undefined;
    }
    if (typeof overrides.gasLimit === 'bigint' ||
        typeof overrides.gasPrice === 'bigint' ||
        typeof overrides.maxFeePerGas === 'bigint' ||
        typeof overrides.maxPriorityFeePerGas === 'bigint') {
        throw new Error('bigint not yet supported');
    }
    return {
        gasLimit: parseNumberOrUndefined(overrides.gasLimit),
        gasPrice: parseHexOrUndefined(overrides.gasPrice),
        maxFeePerGas: parseHexOrUndefined(overrides.maxFeePerGas),
        maxPriorityFeePerGas: parseHexOrUndefined(overrides.maxPriorityFeePerGas),
    };
}
exports.parseTxOverrides = parseTxOverrides;
function parseHexOrUndefined(value) {
    if (typeof value === 'string' && !!value) {
        return value;
    }
    if (typeof value === 'number') {
        return `0x${value.toString(16)}`;
    }
    return undefined;
}
function parseNumberOrUndefined(value) {
    if (typeof value === 'string' && !!value) {
        return Number(value);
    }
    if (typeof value === 'number') {
        return value;
    }
    return undefined;
}
